include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,621 @@
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double dq_relative_error = 0.0;
+
+    /**
+     * The maximum number of beta-condition failures in the
+     * linesearch algorithm. Only used if
+     * strategy==SolutionStrategy::linesearch.
+     */
+    unsigned int maximum_beta_failures;
+
+    /**
+     * The size of the subspace used with Anderson acceleration
+     * in conjunction with Picard or fixed-point iteration.
+     *
+     * If you set this to 0, no acceleration is used.
+     */
+    unsigned int anderson_subspace_size;
+
+    /**
+     * Absolute l2 tolerance of the residual to be reached.
+     *
+     * @note Solver terminates successfully if either the absolute or
+     * the relative tolerance has been reached.
+     */
+    const double absolute_tolerance;
+
+    /**
+     * Relative l2 tolerance of the residual to be reached.
+     *
+     * @note Solver terminates successfully if either the absolute or
+     * the relative tolerance has been reached.
+     */
+    const double relative_tolerance = 1.e-5;
+
+    /**
+     * Number of nonlinear iterations after which the preconditioner
+     * should be updated.
+     */
+    unsigned int threshold_nonlinear_iterations;
+
+    /**
+     * Max number of linear iterations after which the preconditioner
+     * should be updated. This is only used if
+     * solve_with_jacobian_and_track_n_linear_iterations has been given
+     * a target (i.e., it is not empty).
+     */
+    unsigned int threshold_n_linear_iterations;
+
+    /**
+     * Reuse nonlinear solver the next time solve() is called. In particular,
+     * this parameter allows to reuse the preconditioner from the last
+     * solution step, enabling preconditioner lagging over multiple nonlinear
+     * solutions.
+     */
+    const bool reuse_solver = false;
+  };
+
+  /**
+   * Constructor, filling in default values
+   */
+  NonlinearSolverSelector();
+
+  /**
+   * Constructor, selecting the solver @p additional_data.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data);
+
+  /**
+   * Select a new nonlinear solver. All solver names used in this class are
+   * all lower case.
+   */
+  void
+  select(const std::string &type);
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+  void
+  set_data(
+    const typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+      &                                         additional_data,
+    const Teuchos::RCP<Teuchos::ParameterList> &parameters);
+#endif
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_SUNDIALS
+  void
+  set_data(const typename SUNDIALS::KINSOL<VectorType>::AdditionalData
+             &additional_data);
+#endif
+
+  /**
+   * Solve the non linear system. Return the number of nonlinear steps taken
```

Comment:
```suggestion
   * Solve the nonlinear system. Return the number of nonlinear steps taken
```

Reply:
You are returning ``void`` but claim returning the number of iterations. Which one is it?